Native TPC-H Database Generator

bluecat1024 opened the pull request:
native generation for TPC-H workload

apavlo commented on src/main/java/com/oltpbenchmark/benchmarks/tpch/TPCHLoader.java:
```diff
@@ -213,21 +168,11 @@ public void afterLoad() {
-                    LOG.info("Skipping LOAD of Part because COPY has been done.");
-                    return;
-                }
                 try (PreparedStatement statement = conn.prepareStatement("INSERT INTO part " + "(p_partkey, p_name, p_mfgr, p_brand, p_type," + " p_size, p_container, p_retailprice, p_comment) " + "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)")) {
```
Use `SQLUtil.getInsertSQL`

apavlo commented on src/main/java/com/oltpbenchmark/benchmarks/tpch/TPCHLoader.java:
```diff
@@ -169,38 +146,16 @@ private boolean loadCopy(Connection conn) {
-                    copyLatch.await();
-                } catch (InterruptedException e) {
-                    throw new RuntimeException(e);
+                    genTable(conn, statement, regionGenerators, regionTypes, "Region");
```
Don't use hardcoded strings for table names. Define them in `TPCHConstants`.

https://github.com/cmu-db/benchbase/blob/main/src/main/java/com/oltpbenchmark/benchmarks/tatp/TATPConstants.java#L39-L52

apavlo commented on src/main/java/com/oltpbenchmark/benchmarks/tpch/TPCHLoader.java:
```diff
@@ -240,19 +185,17 @@ public void afterLoad() {
-                    LOG.info("Skipping LOAD of Nation because COPY has been done.");
-                    return;
-                }
                 try (PreparedStatement statement = conn.prepareStatement("INSERT INTO nation " + "(n_nationkey, n_name, n_regionkey, n_comment) " + "VALUES (?, ?, ?, ?)")) {
```
Use `SQLUtil.getInsertSQL`

apavlo commented on src/main/java/com/oltpbenchmark/benchmarks/tpch/TPCHLoader.java:
```diff
@@ -268,19 +211,17 @@ public void afterLoad() {
-                    LOG.info("Skipping LOAD of Supplier because COPY has been done.");
-                    return;
-                }
                 try (PreparedStatement statement = conn.prepareStatement("INSERT INTO supplier " + "(s_suppkey, s_name, s_address, s_nationkey, s_phone," + " s_acctbal, s_comment) " + "VALUES (?, ?, ?, ?, ?, ?, ?)")) {
```
Use `SQLUtil.getInsertSQL`

apavlo commented on src/main/java/com/oltpbenchmark/benchmarks/tpch/TPCHLoader.java:
```diff
@@ -296,19 +237,17 @@ public void afterLoad() {
-                    LOG.info("Skipping LOAD of Customer because COPY has been done.");
-                    return;
-                }
                 try (PreparedStatement statement = conn.prepareStatement("INSERT INTO customer " + "(c_custkey, c_name, c_address, c_nationkey," + " c_phone, c_acctbal, c_mktsegment, c_comment ) " + "VALUES (?, ?, ?, ?, ?, ?, ?, ?)")) {
```
Use `SQLUtil.getInsertSQL`

apavlo commented on src/main/java/com/oltpbenchmark/benchmarks/tpch/TPCHLoader.java:
```diff
@@ -324,19 +263,17 @@ public void afterLoad() {
-                    LOG.info("Skipping LOAD of Orders because COPY has been done.");
-                    return;
-                }
                 try (PreparedStatement statement = conn.prepareStatement("INSERT INTO orders " + "(o_orderkey, o_custkey, o_orderstatus, o_totalprice," + " o_orderdate, o_orderpriority, o_clerk, o_shippriority," + " o_comment) " + "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)")) {
```
Use `SQLUtil.getInsertSQL`

apavlo commented on src/main/java/com/oltpbenchmark/benchmarks/tpch/TPCHLoader.java:
```diff
@@ -352,19 +289,17 @@ public void afterLoad() {
-                    LOG.info("Skipping LOAD of PartSupp because COPY has been done.");
-                    return;
-                }
                 try (PreparedStatement statement = conn.prepareStatement("INSERT INTO partsupp " + "(ps_partkey, ps_suppkey, ps_availqty, ps_supplycost," + " ps_comment) " + "VALUES (?, ?, ?, ?, ?)")) {
```
Use `SQLUtil.getInsertSQL`

apavlo commented on src/main/java/com/oltpbenchmark/benchmarks/tpch/TPCHLoader.java:
```diff
@@ -381,19 +316,17 @@ public void afterLoad() {
-                    LOG.info("Skipping LOAD of LineItem because COPY has been done.");
-                    return;
-                }
                 try (PreparedStatement statement = conn.prepareStatement("INSERT INTO lineitem " + "(l_orderkey, l_partkey, l_suppkey, l_linenumber," + " l_quantity, l_extendedprice, l_discount, l_tax," + " l_returnflag, l_linestatus, l_shipdate, l_commitdate," + " l_receiptdate, l_shipinstruct, l_shipmode, l_comment) " + "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)")) {
```
Use `SQLUtil.getInsertSQL`

apavlo commented on src/main/java/com/oltpbenchmark/benchmarks/tpch/generation/AbstractRandomInt.java:
```diff
@@ -0,0 +1,107 @@
+ */
+package com.oltpbenchmark.benchmarks.tpch.generation;
+
+import static com.google.common.base.Preconditions.checkState;
```
Does any other part of BenchBase use this? If no, we should remove it.

apavlo commented on src/main/java/com/oltpbenchmark/benchmarks/tpch/generation/AbstractRandomInt.java:
```diff
@@ -0,0 +1,107 @@
+/*
```
If this code came from the Trino, add their license info to make it clear we got it from them. Do this for all the files that we copy.

apavlo commented on src/main/java/com/oltpbenchmark/benchmarks/tpch/generation/AbstractRandomInt.java:
```diff
@@ -0,0 +1,107 @@
+
+import static com.google.common.base.Preconditions.checkState;
+
+public abstract class AbstractRandomInt
+{
```
Make sure the new code matches the same code style as the rest of the repository.

apavlo commented on src/main/java/com/oltpbenchmark/benchmarks/tpch/generation/Distribution.java:
```diff
@@ -0,0 +1,112 @@
+import java.util.Map;
+import java.util.Map.Entry;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkState;
```
We should probably remove this.

bluecat1024 replied:
I am not sure I have resolved this. Do you mean only the checkState and toStringHelper should be removed, or all com.google.common should be avoided?

apavlo replied:
Do we already include `com.google.common` packages in maven? I don't want additional dependencies just for these checks.

bluecat1024 replied:
Not changing the pom.xml. The repo compiles.

The pom has a com.google.cloud pakage which fetches all google.common so no additional dependency needed.
Yet strangely, google packages are used nowhere except for the TPCH code I have introduced.

bluecat1024 replied:
So the google package functionalities may be replaced by self-defined ones, and com.google.cloud can be instead deleted (google dependency tree is big).

bluecat1024 replied:
resolved this. no google pkg needed.

apavlo commented on src/main/resources/com/oltpbenchmark/benchmarks/tpch/generation/dists.dss:
```diff
@@ -0,0 +1,836 @@
+#
```
@timveil Should this file be in `src/main/resources/benchmarks/tpch`?

bluecat1024 replied:
Shall this be re-located?

apavlo replied:
Yes.

bluecat1024 replied:
resolved

apavlo commented on src/test/java/com/oltpbenchmark/benchmarks/tpch/generation/TestGenerators.java:
```diff
@@ -0,0 +1,95 @@
+
+import com.oltpbenchmark.benchmarks.tpch.generation.*;
+
+public class TestGenerators extends TestCase {
```
Rename to `TestTPCHGenerators`

apavlo commented on src/test/java/com/oltpbenchmark/benchmarks/tpch/generation/TestGenerators.java:
```diff
@@ -0,0 +1,95 @@
+        super();
+    }
+
+    public void testGenerators() {
```
Are you just printing the output of the generators? How are we going to check whether the generators contain correct data when the test runs automatically in Github?

Also if the default scalefactor is one, does this mean that we're going to generate 1GB of debug output?

bluecat1024 replied:
For automatic check I should use assertEqual with some ground truth data.

In those output I hardcoded those SFs to be very small. SFs are parameters for generators.

apavlo replied:
There is already testcase infrastructure for automatically testing benchmark data loaders. Make a new class `TestTPCHLoader` that extends `AbstractTestLoader`:

Examples:
* https://github.com/cmu-db/benchbase/blob/main/src/test/java/com/oltpbenchmark/benchmarks/tpcc/TestTPCCLoader.java
* https://github.com/cmu-db/benchbase/blob/main/src/test/java/com/oltpbenchmark/benchmarks/tatp/TestTATPLoader.java
* thttps://github.com/cmu-db/benchbase/blob/main/src/test/java/com/oltpbenchmark/benchmarks/tpcc/TestTPCCLoader.java

apavlo reviewed (changes requested):
Fix the addressed comments.
Rename `tpch/generation` -> `tpch/util` to match other benchmarks.

Instead of a custom random data generators, I would like to maybe try to combine what we already have with `benchbase/src/main/java/com/oltpbenchmark/util/RandomGenerator.java`. The TPC-H random generator is more advanced, so we should merge their code into our system. But we can will want to get rid of the `expectedUsagePerRow` hint. We use `RandomGenerator` to create input parameters for txns so there isn't a notion of a row in that scenario.

bluecat1024 commented:
> Fix the addressed comments. Rename `tpch/generation` -> `tpch/util` to match other benchmarks.
> 
> Instead of a custom random data generators, I would like to maybe try to combine what we already have with `benchbase/src/main/java/com/oltpbenchmark/util/RandomGenerator.java`. The TPC-H random generator is more advanced, so we should merge their code into our system. But we can will want to get rid of the `expectedUsagePerRow` hint. We use `RandomGenerator` to create input parameters for txns so there isn't a notion of a row in that scenario.

bluecat1024 commented:
> Fix the addressed comments. Rename `tpch/generation` -> `tpch/util` to match other benchmarks.
> 
> Instead of a custom random data generators, I would like to maybe try to combine what we already have with `benchbase/src/main/java/com/oltpbenchmark/util/RandomGenerator.java`. The TPC-H random generator is more advanced, so we should merge their code into our system. But we can will want to get rid of the `expectedUsagePerRow` hint. We use `RandomGenerator` to create input parameters for txns so there isn't a notion of a row in that scenario.

The `expectedUsagePerRow` is a concept for RandomGenerators to skip a certain number of rows and columns, which is often used the native generation of TPCH and TPCDS. So I suppose this part of logic can be refactored as another "RandomInt" in the util package, so both loaders share the logic of it.
